### PR TITLE
fix: use meta device in disable_fp8 to avoid VRAM spike

### LINKED ISSUE #592: Potential VRAM spike (OOM risk) during evaluation in disable_fp8
Hi there! I was studying the code and I found something unusual to me regarding how VRAM is handled during evaluation.

While looking at the `disable_fp8` context manager, I noticed that when it swaps `Float8Linear` back to standard `Linear` layers, it instantiates the new layer using `device=fp8_module.weight.device`. Because it targets the GPU directly rather than using a `meta` device, PyTorch physically allocates a brand new, uninitialized tensor in VRAM for every layer before immediately dropping the reference.

I wanted to be absolutely sure, so I wrote a test script to track `torch.cuda.max_memory_allocated()`. On a dummy model with `n_embd=8192`, running this context manager caused a sudden 1 GB memory spike. If a user is pushing their `device-batch-size` right up to their VRAM limit, this could trigger an unexpected OOM crash during the evaluation phase.

### To Reproduce
Here is the output from my test tracking peak memory on the original code during the context manager setup:
```text
1. Initializing real GPT model on meta device...
2. Moving model to GPU and initializing weights...
3. Running real FP8 conversion...

=> Base VRAM Usage (Model Size): 8197.00 MB

--- Running ORIGINAL Buggy Context Manager ---
Peak VRAM during eval setup: 9221.00 MB
Unnecessary Memory Spike:  + 1024.00 MB

--- Running FIXED Context Manager ---
Peak VRAM during eval setup: 8197.01 MB
Unnecessary Memory Spike:  + 0.01 MB

--- Running Accuracy & Memory Verification ---
Original Memory Address: 140317353312256
Swapped Memory Address:  140317353312256
Original Weight Sum:     152.7218
Swapped Weight Sum:      152.7218
SUCCESS: The fixed context manager uses the exact same physical memory and math!
```

Please let me know if I'm missing some context and this temporary VRAM allocation is actually intended behavior, I'm still learning the codebase!

```python

import torch
from contextlib import contextmanager

# 1. IMPORT ACTUAL MODULES FROM YOUR CODEBASE
from nanochat.gpt import GPT, GPTConfig, Linear
from nanochat.fp8 import convert_to_float8_training, Float8LinearConfig

# ==========================================
# 2. THE EXACT CONTEXT MANAGER FROM base_train.py
# ==========================================
@contextmanager
def disable_fp8_buggy(model):
    import torch.nn as nn

    fp8_locations = []
    for name, module in model.named_modules():
        if 'Float8' in type(module).__name__:
            if '.' in name:
                parent_name, attr_name = name.rsplit('.', 1)
                parent = model.get_submodule(parent_name)
            else:
                parent = model
                attr_name = name
            fp8_locations.append((parent, attr_name, module))

    if not fp8_locations:
        yield
        return

    for parent, attr_name, fp8_module in fp8_locations:
        # THE BUG IS HERE: device=cuda causes immediate VRAM allocation
        linear = Linear(
            fp8_module.in_features,
            fp8_module.out_features,
            bias=fp8_module.bias is not None,
            device=fp8_module.weight.device,
            dtype=fp8_module.weight.dtype,
        )
        linear.weight = fp8_module.weight  # Original tensor orphaned here
        if fp8_module.bias is not None:
            linear.bias = fp8_module.bias
        setattr(parent, attr_name, linear)

    try:
        yield
    finally:
        for parent, attr_name, fp8_module in fp8_locations:
            setattr(parent, attr_name, fp8_module)

# ==========================================
# 3. THE FIXED CONTEXT MANAGER (Using Meta)
# ==========================================
@contextmanager
def disable_fp8_fixed(model):
    import torch.nn as nn

    fp8_locations = []
    for name, module in model.named_modules():
        if 'Float8' in type(module).__name__:
            if '.' in name:
                parent_name, attr_name = name.rsplit('.', 1)
                parent = model.get_submodule(parent_name)
            else:
                parent = model
                attr_name = name
            fp8_locations.append((parent, attr_name, module))

    if not fp8_locations:
        yield
        return

    for parent, attr_name, fp8_module in fp8_locations:
        # THE FIX: Wrap in meta device to prevent VRAM allocation
        with torch.device("meta"):
            linear = Linear(
                fp8_module.in_features,
                fp8_module.out_features,
                bias=fp8_module.bias is not None,
                dtype=fp8_module.weight.dtype,
            )
        linear.weight = fp8_module.weight
        if fp8_module.bias is not None:
            linear.bias = fp8_module.bias
        setattr(parent, attr_name, linear)

    try:
        yield
    finally:
        for parent, attr_name, fp8_module in fp8_locations:
            setattr(parent, attr_name, fp8_module)

# ==========================================
# 4. THE EXPERIMENT
# ==========================================
def run_experiment():
    if not torch.cuda.is_available():
        print("This script requires a CUDA-enabled GPU.")
        return

    device = torch.device("cuda:0")

    config = GPTConfig(
        n_layer=2,          
        n_embd=8192,        
        n_head=32,
        n_kv_head=32,
        vocab_size=32768,
        sequence_len=1024
    )

    print("1. Initializing real GPT model on meta device...")
    with torch.device("meta"):
        model = GPT(config)
    
    print("2. Moving model to GPU and initializing weights...")
    model.to_empty(device=device)
    model.init_weights()

    print("3. Running real FP8 conversion...")
    def fp8_filter(mod, fqn):
        import torch.nn as nn
        if not isinstance(mod, nn.Linear): return False
        if mod.in_features % 16 != 0 or mod.out_features % 16 != 0: return False
        if min(mod.in_features, mod.out_features) < 128: return False
        return True

    fp8_config = Float8LinearConfig.from_recipe_name("tensorwise")
    convert_to_float8_training(model, config=fp8_config, module_filter_fn=fp8_filter)

    base_mem = torch.cuda.memory_allocated() / (1024 ** 2)
    print(f"\n=> Base VRAM Usage (Model Size): {base_mem:.2f} MB")

    # --- TEST 1: BUGGY ---
    torch.cuda.reset_peak_memory_stats()
    print("\n--- Running ORIGINAL Buggy Context Manager ---")
    with disable_fp8_buggy(model):
        peak_mem = torch.cuda.max_memory_allocated() / (1024 ** 2)
        spike = peak_mem - base_mem
        print(f"Peak VRAM during eval setup: {peak_mem:.2f} MB")
        print(f"Unnecessary Memory Spike:  + {spike:.2f} MB")

    # --- TEST 2: FIXED WITH ACCURACY CHECK ---
    torch.cuda.reset_peak_memory_stats()
    print("\n--- Running FIXED Context Manager ---")
    
    # Grab the data pointer and the mathematical sum of the weights BEFORE the context manager
    # We use a massive layer (c_fc in the MLP) for the test
    target_layer = model.transformer.h[0].mlp.c_fc
    original_ptr = target_layer.weight.data_ptr()
    original_sum = target_layer.weight.float().sum().item()
    
    with disable_fp8_fixed(model):
        peak_mem = torch.cuda.max_memory_allocated() / (1024 ** 2)
        spike = peak_mem - base_mem
        print(f"Peak VRAM during eval setup: {peak_mem:.2f} MB")
        print(f"Unnecessary Memory Spike:  + {spike:.2f} MB")
        
        print("\n--- Running Accuracy & Memory Verification ---")
        # Grab the data pointer and the mathematical sum of the weights INSIDE the context manager
        # (Where the module has been swapped to 'Linear')
        swapped_layer = model.transformer.h[0].mlp.c_fc
        new_ptr = swapped_layer.weight.data_ptr()
        new_sum = swapped_layer.weight.float().sum().item()
        
        print(f"Original Memory Address: {original_ptr}")
        print(f"Swapped Memory Address:  {new_ptr}")
        print(f"Original Weight Sum:     {original_sum:.4f}")
        print(f"Swapped Weight Sum:      {new_sum:.4f}")
        
        if original_ptr == new_ptr and original_sum == new_sum:
            print("SUCCESS: The fixed context manager uses the exact same physical memory and math!")
        else:
            print("ERROR: Mismatch detected!")

if __name__ == "__main__":
    run_experiment()
```

### PR DESCRIPTION
**Pull request by Jah-yee**
Good day!

This PR fixes the VRAM spike issue in the `disable_fp8` context manager (issue #592).

## Problem
When swapping Float8Linear to Linear, using `device=fp8_module.weight.device` directly allocates new tensors on GPU, causing unnecessary VRAM spike (~1GB for large models). This can trigger unexpected OOM crashes during evaluation for users pushing their batch size to VRAM limits.

## Solution
Use `device="meta"` to avoid physical memory allocation, then swap in the weight tensor reference. This eliminates the unnecessary VRAM spike while maintaining identical functionality and numerical accuracy.

## Testing
The fix has been verified to:
- Eliminate the VRAM spike (from ~1GB to ~0MB)
- Maintain identical memory addresses for weight tensors
- Preserve numerical accuracy (weight sums match)

Thank you for your work on this project. I hope this fix can help. Please let me know if there are any issues or suggestions, and I will address them.

Warmly,
RoomWithOutRoof

### CODE REVIEW

**Comment by karpathy** [comment]
Thank you for the PR you're right.
